internal/jimm/bootstrap/bootstrap_job.go

```diff
@@ -0,0 +1,245 @@
+	RunWrapper(
+		ctx context.Context,
+		binaryPath, jujuDataDir string,
+		params jujucommands.BootstrapCmdParams,
+	) (<-chan jujucommands.OutputLine, jujuclient.ClientStore, func(), error)
+}
+
+// DefaultBootstrapExecutor defines the default, and expected bootstrap executor
+// to be used in a [bootstrapManager.BootstrapJob]
+type DefaultBootstrapExecutor struct{}
+
+// RunWrapper wraps the command runner and bootstrap command to be run, and then runs it for you.
+// This enables the running portion of the BootstrapJob to be mocked.
+func (h DefaultBootstrapExecutor) RunWrapper(
+	ctx context.Context,
+	binaryPath, jujuDataDir string,
+	params jujucommands.BootstrapCmdParams,
+) (<-chan jujucommands.OutputLine, jujuclient.ClientStore, func(), error) {
+	r := jujucommands.NewCommandRunner(binaryPath, jujuDataDir)
+	command := jujucommands.NewBootstrapCmd(r)
+	return command.Run(ctx, params)
+}
+
+// JobParams holds the params to run a juju bootstrap job.
+type JobParams struct {
+	// Runner params.
+
+	JujuDataDir string
+
+	// CLI Download params.
+
+	CLIVersion string
+	CLIOs      string
+	CLIArch    string
+
+	// User defined command arguments
+
+	CloudNameAndRegion string
+	ControllerName     string
+	AgentVersion       string
+	BootstrapTimeout   int
+	CloudCred          jujucloud.CloudCredential
+	// PersonalCloud is the personally defined cloud. Only necessary if the cloud is not a public
+	// cloud.
+	PersonalCloud jujucloud.Cloud
+
+	// JIMM Provided command arguments (i.e., ones that must be set by JIMM when bootstrapping).
+
+	LoginTokenRefreshURL string
+}
+
+// BootstrapJob returns a [jobtracker.JobFunc] [for use in the [jobtracker.Tracker]] responsible for
+// bootstrapping a controller and adding it to JIMM.
+func (b *bootstrapManager) BootstrapJob(
+	p JobParams,
+	executor BootstrapExecutor,
+	user *openfga.User,
+) jobtracker.JobFunc {
+	const bootstrapLockTTL = 40 * time.Minute
+
+	return func(ctx context.Context) error {
+		jobId, ok := jobtracker.JobIdFromContext(ctx)
+		if !ok {
+			return fmt.Errorf("failed to get job ID from context")
+		}
+
+		zapctx.Debug(
+			ctx,
+			"starting bootstrap job",
+			zap.String("job-id", jobId.String()),
+			zap.String("controller-name", p.ControllerName),
+		)
+
+		if err := b.store.LockBootstrap(ctx, bootstrapLockTTL); err != nil {
+			zapctx.Error(
+				ctx,
+				"failed to acquire bootstrap lock",
+				zap.String("job-id", jobId.String()),
+				zap.String("controller-name", p.ControllerName),
+				zap.Error(err),
+			)
+			return fmt.Errorf("failed to acquire bootstrap lock: %w", err)
+		}
+
+		defer func() {
+			if err := b.store.UnlockBootstrap(ctx); err != nil {
+				zapctx.Error(
+					ctx,
+					"failed to unlock bootstrap lock",
+					zap.String("job-id", jobId.String()),
+					zap.String("controller-name", p.ControllerName),
+					zap.Error(err),
+				)
+			}
+		}()
+
+		err := b.store.GetController(ctx, &dbmodel.Controller{Name: p.ControllerName})
```

Comment:
This check isn't strong enough to prevent someone from bootstrapping twice with the same controller name. I know our bootstrap is limited to 1 controller at a time, but if that wasn't the case then you can have 2 API calls to bootstrap pass this check. I don't think there's a simple solution though because creating a controller in the DB is not currently designed for mid-bootstrap controllers.

Reply:
just create an "empty" controller in the DB... we should have an index on controller name, so it'll fail if somebody tries to bootstrap with the same name, no?